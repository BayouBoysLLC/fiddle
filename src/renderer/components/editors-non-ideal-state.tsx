import * as React from 'react';

import { Button, NonIdealState } from '@blueprintjs/core';
import { Fiddle } from '../fiddle';

export function renderNonIdealState(fiddle: Fiddle) {
  const resolveButton = (
    <Button text="Open all editors" onClick={() => fiddle.showAll()} />
  );

  return (
    <NonIdealState
      action={resolveButton}
      icon="applications"
      description='You have closed all editors. You can open them again with the button below or the "Editors" button above!'
    />
  );
}
