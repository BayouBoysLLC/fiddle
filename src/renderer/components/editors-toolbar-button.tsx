import * as React from 'react';

import { Button } from '@blueprintjs/core';
import {
  MosaicContext,
  MosaicRootActions,
  MosaicWindowContext,
} from 'react-mosaic-component';

import { EditorId } from '../../interfaces';
import { Fiddle } from '../fiddle';

interface ToolbarButtonProps {
  fiddle: Fiddle;
  id: EditorId;
}

abstract class ToolbarButton extends React.PureComponent<ToolbarButtonProps> {
  public static contextType = MosaicWindowContext;
  public context: MosaicWindowContext;

  constructor(props: ToolbarButtonProps) {
    super(props);
  }

  public render() {
    return (
      <MosaicContext.Consumer>
        {({ mosaicActions }) => this.createButton(mosaicActions)}
      </MosaicContext.Consumer>
    );
  }

  /**
   * Create a button that performs the actual action
   */
  public abstract createButton(
    _mosaicActions: MosaicRootActions<any>,
  ): React.ReactNode;
}

export class MaximizeButton extends ToolbarButton {
  /**
   * Create a button that can expand this panel
   */
  public createButton(mosaicActions: MosaicRootActions<any>) {
    const onClick = () => {
      mosaicActions.expand(this.context.mosaicWindowActions.getPath());
    };

    return <Button icon="maximize" className="bp3-small" onClick={onClick} />;
  }
}

export class RemoveButton extends ToolbarButton {
  /**
   * Create a button that can remove this panel
   */
  public createButton(_mosaicActions: MosaicRootActions<any>) {
    // FIXME(ckerr) is this button supposed to remove or to hide
    const onClick = () => this.props.fiddle.hide(this.props.id);

    return <Button icon="cross" className="bp3-small" onClick={onClick} />;
  }
}
