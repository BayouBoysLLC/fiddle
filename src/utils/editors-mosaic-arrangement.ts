import { MosaicDirection, MosaicNode, getLeaves } from 'react-mosaic-component';

import { EditorId } from '../interfaces';

/**
 * Create a mosaic arrangement given an array of editor ids.
 *
 * @export
 * @param {Array<EditorId>} input
 * @returns {MosaicNode<EditorId>}
 */
export function createMosaicArrangement(
  input: readonly EditorId[],
  direction: MosaicDirection = 'row',
): MosaicNode<EditorId> {
  if (input.length === 1) {
    return input[0];
  }

  // This cuts out the first half of input. Input becomes the second half.
  const secondHalf = [...input];
  const firstHalf = secondHalf.splice(0, Math.floor(secondHalf.length / 2));

  return {
    direction,
    first: createMosaicArrangement(firstHalf, 'column'),
    second: createMosaicArrangement(secondHalf, 'column'),
  };
}

/**
 * Returns an array of visible editors given a mosaic arrangement.
 *
 * @export
 * @param {MosaicNode<EditorId> | null} input
 * @returns {Array<EditorId>}
 */
export function getVisibleMosaics(
  input: MosaicNode<EditorId> | null,
): Array<EditorId> {
  return getLeaves(input);
}
