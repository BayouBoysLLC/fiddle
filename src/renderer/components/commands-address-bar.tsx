import { Button, InputGroup, Intent } from '@blueprintjs/core';
import classnames from 'classnames';
import { reaction } from 'mobx';
import { observer } from 'mobx-react';
import * as React from 'react';

import { AppState } from '../state';
import { EditorMosaic } from '../editor-mosaic';
import { GistActionState } from '../../interfaces';
import { IpcEvents } from '../../ipc-events';
import { getGistId, urlFromId } from '../../utils/gist';
import { ipcRendererManager } from '../ipc';

interface AddressBarProps {
  appState: AppState;
  editorMosaic: EditorMosaic;
}

interface AddressBarState {
  value: string;
  loaders: {
    gist: any;
    example: any;
  };
}

@observer
export class AddressBar extends React.Component<
  AddressBarProps,
  AddressBarState
> {
  constructor(props: AddressBarProps) {
    super(props);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
    this.submit = this.submit.bind(this);

    const { gistId } = this.props.appState;
    const value = urlFromId(gistId);

    const { remoteLoader } = window.ElectronFiddle.app;

    this.state = {
      value,
      loaders: {
        gist: remoteLoader.loadFiddleFromGist.bind(remoteLoader),
        example: remoteLoader.loadFiddleFromElectronExample.bind(remoteLoader),
      },
    };
  }

  /**
   * Handle the form's submit event, trying to load whatever
   * URL was entered.
   *
   * @param {React.SyntheticEvent<HTMLFormElement>} event
   */
  public handleSubmit(event: React.SyntheticEvent<HTMLFormElement>) {
    event.preventDefault();
    this.submit();
  }

  /**
   * Commit the address bar's value to app state and load the fiddle.
   *
   * @memberof AddressBar
   */
  public submit() {
    const gistId = getGistId(this.state.value);
    window.ElectronFiddle.app.openFiddle({ gistId });
  }

  /**
   * Once the component mounts, we'll subscribe to gistId changes
   */
  public componentDidMount() {
    const { appState } = this.props;
    const { loaders } = this.state;
    reaction(
      () => appState.gistId,
      (gistId: string) => this.setState({ value: urlFromId(gistId) }),
    );
    ipcRendererManager.on(IpcEvents.LOAD_GIST_REQUEST, loaders.gist);
    ipcRendererManager.on(
      IpcEvents.LOAD_ELECTRON_EXAMPLE_REQUEST,
      loaders.example,
    );
  }

  public componentWillUnmount() {
    const { loaders } = this.state;
    ipcRendererManager.removeListener(
      IpcEvents.LOAD_GIST_REQUEST,
      loaders.gist,
    );
    ipcRendererManager.removeListener(
      IpcEvents.LOAD_ELECTRON_EXAMPLE_REQUEST,
      loaders.example,
    );
  }

  /**
   * Handle the change event, which usually just updates the address bar's value
   *
   * @param {React.ChangeEvent<HTMLInputElement>} event
   */
  public handleChange(event: React.ChangeEvent<HTMLInputElement>) {
    this.setState({ value: event.target.value });
  }

  /**
   * If the input text is empty when it loses focus,
   * reset it to the current gist
   */
  public handleBlur(event: React.FocusEvent<HTMLInputElement>) {
    if (!event.target.value) {
      this.setState({ value: urlFromId(this.props.appState.gistId) });
    }
  }

  public renderLoadButton(isValueCorrect: boolean): JSX.Element {
    return (
      <Button
        disabled={!isValueCorrect}
        icon="cloud-download"
        text="Load Fiddle"
        onClick={this.submit}
      />
    );
  }

  public render() {
    const { activeGistAction } = this.props.appState;
    const { isEdited } = this.props.editorMosaic;
    const { value } = this.state;
    const isCorrect = !!getGistId(value);
    const className = classnames('address-bar', isEdited, { empty: !value });

    const isPerformingAction = activeGistAction !== GistActionState.none;
    return (
      <form className={className} onSubmit={this.handleSubmit}>
        <fieldset disabled={isPerformingAction}>
          <InputGroup
            key="addressbar"
            leftIcon="geosearch"
            intent={isCorrect || !value ? undefined : Intent.DANGER}
            onChange={this.handleChange}
            onBlur={this.handleBlur}
            placeholder="https://gist.github.com/..."
            value={value}
            rightElement={this.renderLoadButton(isCorrect)}
          />
        </fieldset>
      </form>
    );
  }
}
