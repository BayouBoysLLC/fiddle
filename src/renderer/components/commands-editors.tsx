import {
  Button,
  Menu,
  MenuDivider,
  MenuItem,
  Popover,
  Position,
} from '@blueprintjs/core';
import { when } from 'mobx';
import { observer } from 'mobx-react';
import * as React from 'react';

import { EditorId, GenericDialogType, MAIN_JS } from '../../interfaces';
import { getEditorTitle, getEmptyContent } from '../../utils/editor-utils';
import { AppState } from '../state';
import { EditorState, Fiddle } from '../fiddle';

interface EditorDropdownState {
  value: string;
}

interface EditorDropdownProps {
  appState: AppState;
  fiddle: Fiddle;
}

/**
 * A dropdown allowing users to toggle the various editors
 *
 * @class EditorDropdown
 * @extends {React.Component<EditorDropdownProps, EditorDropdownState>}
 */
@observer
export class EditorDropdown extends React.Component<
  EditorDropdownProps,
  EditorDropdownState
> {
  constructor(props: EditorDropdownProps) {
    super(props);

    this.addEditor = this.addEditor.bind(this);
    this.onItemClick = this.onItemClick.bind(this);
    this.removeEditor = this.removeEditor.bind(this);
    this.showEditorDialog = this.showEditorDialog.bind(this);
  }

  public render() {
    return (
      <>
        <Popover content={this.renderMenu()} position={Position.BOTTOM}>
          <Button icon="applications" text="Editors" />
        </Popover>
      </>
    );
  }

  public renderMenu() {
    return <Menu>{...this.renderMenuItems()}</Menu>;
  }

  public renderMenuItems() {
    const { fiddle } = this.props;
    const { mosaicLeafCount, states } = fiddle;
    const result: Array<JSX.Element> = [];

    for (const [id, state] of states) {
      const visible =
        state === EditorState.Visible || state === EditorState.Pending;
      const icon = visible ? 'eye-open' : 'eye-off';
      const title = getEditorTitle(id);
      const mustShow = visible && mosaicLeafCount < 2;

      if (id !== MAIN_JS) {
        result.push(
          <MenuItem
            disabled={mustShow}
            icon={icon}
            id={id}
            key={id}
            onClick={this.onItemClick}
            text={title}
          >
            <MenuItem
              icon={'cross'}
              id={id}
              onClick={this.removeEditor}
              text={'Remove'}
            />
          </MenuItem>,
        );
      } else {
        result.push(
          <MenuItem
            disabled={mustShow}
            icon={icon}
            id={id}
            key={id}
            onClick={this.onItemClick}
            text={title}
          />,
        );
      }
    }

    result.push(
      <React.Fragment key={'fragment-editor'}>
        <MenuDivider />
        <MenuItem
          icon="plus"
          key="add-editor"
          text="Add Editor"
          onClick={this.addEditor}
        />
      </React.Fragment>,
    );

    result.push(
      <React.Fragment key={'fragment-reset-layout'}>
        <MenuDivider />
        <MenuItem
          icon="grid-view"
          key="reset-layout"
          text="Reset Layout"
          onClick={fiddle.resetLayout}
        />
      </React.Fragment>,
    );

    return result;
  }

  public async showEditorDialog() {
    const { appState } = this.props;

    appState.setGenericDialogOptions({
      type: GenericDialogType.confirm,
      label: 'Enter a filename to add',
      wantsInput: true,
      ok: 'Create',
      cancel: 'Cancel',
      placeholder: 'file.js',
    });

    appState.toggleGenericDialog();
    await when(() => !appState.isGenericDialogShowing);

    return {
      cancelled: !appState.genericDialogLastResult,
      result: appState.genericDialogLastInput,
    };
  }

  public async addEditor() {
    const { appState, fiddle } = this.props;

    const { cancelled, result } = await this.showEditorDialog();
    if (cancelled) return;

    try {
      const id = result as EditorId;
      fiddle.add(id, getEmptyContent(id));
    } catch (error) {
      appState.setGenericDialogOptions({
        type: GenericDialogType.warning,
        label: error.message,
        cancel: undefined,
      });
      appState.toggleGenericDialog();
    }
  }

  public async removeEditor(event: React.MouseEvent) {
    const { fiddle } = this.props;
    const { id } = event.currentTarget;

    console.log(`EditorDropdown: Removing editor ${id}`);
    fiddle.remove(id as EditorId);
  }

  public onItemClick(event: React.MouseEvent) {
    const { fiddle } = this.props;
    const { id } = event.currentTarget;

    console.log('onItemClick calling fiddle.toggle', id);
    fiddle.toggle(id as EditorId);
  }
}
