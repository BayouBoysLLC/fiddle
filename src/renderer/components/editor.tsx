// Credit goes in large part to https://github.com/superRaytin/react-monaco-editor,
// this component is a changed version of it.

import * as MonacoType from 'monaco-editor';
import * as React from 'react';

import { AppState } from '../state';
import { EditorId } from '../../interfaces';
import { EditorMosaic } from '../editor-mosaic';
import { monacoLanguage } from '../../utils/editor-utils';

interface EditorProps {
  appState: AppState;
  editorMosaic: EditorMosaic;
  monaco: typeof MonacoType;
  monacoOptions: MonacoType.editor.IEditorOptions;
  id: EditorId;
  options?: Partial<MonacoType.editor.IEditorConstructionOptions>;
  editorDidMount?: (editor: MonacoType.editor.IStandaloneCodeEditor) => void;
  onChange?: (
    value: string,
    event: MonacoType.editor.IModelContentChangedEvent,
  ) => void;
  setFocused: (id: EditorId) => void;
}

export class Editor extends React.Component<EditorProps> {
  public editor: MonacoType.editor.IStandaloneCodeEditor;
  public language = 'javascript';
  public value = '';

  private containerRef = React.createRef<HTMLDivElement>();

  constructor(props: EditorProps) {
    super(props);
    this.language = monacoLanguage(props.id);
  }

  public shouldComponentUpdate() {
    return false;
  }

  public async componentDidMount() {
    return this.initMonaco();
  }

  public componentWillUnmount() {
    this.destroyMonaco();
  }

  /**
   * Handle the editor having been mounted. This refers to Monaco's
   * mount, not React's.
   *
   * @param {MonacoType.editor.IStandaloneCodeEditor} editor
   */
  public async editorDidMount(editor: MonacoType.editor.IStandaloneCodeEditor) {
    const { editorMosaic, editorDidMount, id } = this.props;

    // Set the editor as an available object.
    editorMosaic.addEditor(id, editor);

    // And notify others
    if (editorDidMount) {
      editorDidMount(editor);
    }
  }

  /**
   * Initialize Monaco.
   */
  public async initMonaco() {
    const { monaco, monacoOptions: monacoOptions } = this.props;
    const ref = this.containerRef.current;

    if (ref) {
      this.editor = monaco.editor.create(ref, {
        language: this.language,
        theme: 'main',
        contextmenu: false,
        model: null,
        ...monacoOptions,
      });

      // mark this editor as focused whenever it is
      console.log('this.editor', JSON.stringify(this.editor));
      this.editor.onDidFocusEditorText(() => {
        const { id, setFocused } = this.props;
        setFocused(id);
      });

      await this.editorDidMount(this.editor);
    }
  }

  /**
   * Destroy Monaco.
   */
  public destroyMonaco() {
    if (typeof this.editor !== 'undefined') {
      console.log('Editor: Disposing');
      this.editor.dispose();
    }
  }

  public render() {
    return <div className="editorContainer" ref={this.containerRef} />;
  }
}
