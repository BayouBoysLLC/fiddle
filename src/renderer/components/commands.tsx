import { Button, ControlGroup } from '@blueprintjs/core';
import { observer } from 'mobx-react';
import * as React from 'react';

import { AppState } from '../state';
import { AddressBar } from './commands-address-bar';
import { BisectHandler } from './commands-bisect';
import { EditorDropdown } from './commands-editors';
import { Fiddle } from '../fiddle';
import { GistActionButton } from './commands-action-button';
import { Runner } from './commands-runner';
import { VersionChooser } from './commands-version-chooser';

interface CommandsProps {
  appState: AppState;
  fiddle: Fiddle;
}

/**
 * The command bar, containing all the buttons doing
 * all the things
 *
 * @class Commands
 * @extends {React.Component<CommandsProps>}
 */
@observer
export class Commands extends React.Component<CommandsProps> {
  constructor(props: CommandsProps) {
    super(props);
  }

  public render() {
    const { appState, fiddle } = this.props;
    const { isBisectCommandShowing: isBisectCommandShowing } = appState;
    console.log('Commands.render', 'fiddle.visible', fiddle.visible);

    return (
      <div className="commands">
        <div>
          <ControlGroup fill={true} vertical={false}>
            <VersionChooser appState={appState} />
            <Runner appState={appState} />
          </ControlGroup>
          {isBisectCommandShowing && (
            <ControlGroup fill={true} vertical={false}>
              <BisectHandler appState={appState} />
            </ControlGroup>
          )}
          <ControlGroup fill={true} vertical={false}>
            <Button
              active={appState.isConsoleShowing}
              icon="console"
              text="Console"
              onClick={appState.toggleConsole}
            />
            <EditorDropdown appState={appState} fiddle={fiddle} />
          </ControlGroup>
        </div>
        <div>
          <AddressBar appState={appState} fiddle={fiddle} />
          <GistActionButton appState={appState} fiddle={fiddle} />
        </div>
      </div>
    );
  }
}
