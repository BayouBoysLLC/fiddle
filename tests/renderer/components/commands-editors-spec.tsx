import { mount, shallow } from 'enzyme';
import * as React from 'react';

import { EditorId, GenericDialogType, MAIN_JS } from '../../../src/interfaces';
import { EditorDropdown } from '../../../src/renderer/components/commands-editors';
import { getVisibleMosaics } from '../../../src/utils/editors-mosaic-arrangement';

jest.mock('../../../src/utils/editors-mosaic-arrangement');

describe('EditorDropdown component', () => {
  let store: any;
  let initialMosaics: EditorId[];
  let visibleMosaics: EditorId[];

  beforeEach(() => {
    initialMosaics = [MAIN_JS];
    visibleMosaics = [...initialMosaics];
    store = {
      allMosaics: [...initialMosaics],
      closedPanels: {},
      hideAndBackupMosaic: jest.fn().mockImplementation((id: EditorId) => {
        store.closedPanels[id] = true;
        visibleMosaics = visibleMosaics.filter((x: EditorId) => x !== id);
      }),
      removeMosaic: jest.fn().mockImplementation((id: EditorId) => {
        visibleMosaics = visibleMosaics.filter((x: EditorId) => x !== id);
        store.allMosaics = store.allMosaics.filter((x: EditorId) => x !== id);
      }),
      setGenericDialogOptions: jest.fn(),
      showMosaic: jest.fn(),
      toggleGenericDialog: jest.fn(),
    };

    (getVisibleMosaics as jest.Mock).mockImplementation(() => visibleMosaics);
  });

  afterEach(() => {
    store = {
      allMosaics: [],
      closedPanels: {},
      genericDialogLastInput: undefined,
    };
  });

  it('renders', () => {
    const wrapper = shallow(<EditorDropdown appState={store} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('handles a click for an item', () => {
    const id = initialMosaics[0];
    const wrapper = mount(<EditorDropdown appState={store} />);
    const dropdown = wrapper.instance() as EditorDropdown;

    dropdown.onItemClick({ currentTarget: { id } } as any);
    expect(store.hideAndBackupMosaic).toHaveBeenCalledTimes(1);
    expect(store.showMosaic).toHaveBeenCalledTimes(0);

    dropdown.onItemClick({ currentTarget: { id } } as any);
    expect(store.hideAndBackupMosaic).toHaveBeenCalledTimes(1);
    expect(store.showMosaic).toHaveBeenCalledTimes(1);
  });

  it('disables hide button if only one editor open', () => {
    store.mosaicArrangement = MAIN_JS;
    // (getVisibleMosaics as jest.Mock).mockReturnValue([DefaultEditorId.html]);

    const wrapper = mount(<EditorDropdown appState={store} />);
    const instance = wrapper.instance() as EditorDropdown;
    const menu = instance.renderMenuItems();

    expect(menu).toMatchSnapshot();
  });

  it('can add a valid editor', async () => {
    const file = 'file.js';

    store.showEditorDialog = jest
      .fn()
      .mockReturnValue({ cancelled: false, result: file });
    const wrapper = mount(<EditorDropdown appState={store} />);
    const dropdown = wrapper.instance() as EditorDropdown;

    store.genericDialogLastInput = file;
    store.genericDialogLastResult = true;

    await dropdown.addEditor();

    expect(store.setGenericDialogOptions).toHaveBeenNthCalledWith(1, {
      type: GenericDialogType.confirm,
      label: 'Enter a filename to add',
      cancel: 'Cancel',
      placeholder: 'file.js',
      ok: 'Create',
      wantsInput: true,
    });

    expect(store.showMosaic).toHaveBeenCalledTimes(1);
    expect(store.toggleGenericDialog).toHaveBeenCalledTimes(1);
    expect([...store.allMosaics].sort()).toEqual(
      [...initialMosaics, file].sort(),
    );
  });

  it('errors when trying to add a duplicate editor', async () => {
    const dupe = initialMosaics[0];

    store.showEditorDialog = jest
      .fn()
      .mockReturnValue({ cancelled: false, result: dupe });
    const wrapper = mount(<EditorDropdown appState={store} />);
    const dropdown = wrapper.instance() as EditorDropdown;

    store.genericDialogLastInput = dupe;
    store.genericDialogLastResult = true;

    await dropdown.addEditor();

    expect(store.setGenericDialogOptions).toHaveBeenNthCalledWith(1, {
      type: GenericDialogType.confirm,
      label: 'Enter a filename to add',
      cancel: 'Cancel',
      placeholder: 'file.js',
      ok: 'Create',
      wantsInput: true,
    });

    expect(store.setGenericDialogOptions).toHaveBeenNthCalledWith(2, {
      type: GenericDialogType.warning,
      label: `Filename "${dupe}" already exists - duplicates are not allowed`,
      cancel: undefined,
    });

    expect(store.toggleGenericDialog).toHaveBeenCalledTimes(2);
    expect(store.showMosaic).toHaveBeenCalledTimes(0);
    expect(store.allMosaics).toEqual(initialMosaics);
  });

  it('errors when trying to add an invalid editor', async () => {
    const badFile = 'ðŸ’€.ðŸ’©';

    store.showEditorDialog = jest
      .fn()
      .mockReturnValue({ cancelled: false, result: badFile });
    const wrapper = mount(<EditorDropdown appState={store} />);
    const dropdown = wrapper.instance() as EditorDropdown;

    store.genericDialogLastInput = badFile;
    store.genericDialogLastResult = true;

    await dropdown.addEditor();

    expect(store.setGenericDialogOptions).toHaveBeenNthCalledWith(1, {
      cancel: 'Cancel',
      label: 'Enter a filename to add',
      ok: 'Create',
      placeholder: 'file.js',
      type: GenericDialogType.confirm,
      wantsInput: true,
    });

    expect(store.setGenericDialogOptions).toHaveBeenNthCalledWith(2, {
      cancel: undefined,
      label: 'Invalid editor name - must be either an html, js, or css file.',
      type: GenericDialogType.warning,
    });

    expect(store.showMosaic).toHaveBeenCalledTimes(0);
    expect(store.toggleGenericDialog).toHaveBeenCalledTimes(2);
    expect(store.allMosaics).toEqual(initialMosaics);
  });

  it('can remove an editor', () => {
    const file = 'file.js';
    store.allMosaics = [file];

    const wrapper = mount(<EditorDropdown appState={store} />);
    const dropdown = wrapper.instance() as EditorDropdown;

    dropdown.removeMosaic({
      currentTarget: { id: file },
    } as any);

    expect(store.removeMosaic).toHaveBeenCalledTimes(1);
  });
});
