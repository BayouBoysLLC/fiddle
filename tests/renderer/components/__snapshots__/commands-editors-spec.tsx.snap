// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorDropdown component disables hide button if only one editor open 1`] = `
Array [
  <Blueprint3.MenuItem
    disabled={true}
    icon="eye-open"
    id="main.js"
    multiline={false}
    onClick={[Function]}
    popoverProps={Object {}}
    shouldDismissPopover={true}
    text="Main Process (main.js)"
  />,
  <React.Fragment>
    <Blueprint3.MenuDivider />
    <Blueprint3.MenuItem
      disabled={false}
      icon="plus"
      multiline={false}
      onClick={[Function]}
      popoverProps={Object {}}
      shouldDismissPopover={true}
      text="Add Editor"
    />
  </React.Fragment>,
  <React.Fragment>
    <Blueprint3.MenuDivider />
    <Blueprint3.MenuItem
      disabled={false}
      icon="grid-view"
      multiline={false}
      onClick={[Function]}
      popoverProps={Object {}}
      shouldDismissPopover={true}
      text="Reset Layout"
    />
  </React.Fragment>,
]
`;

exports[`EditorDropdown component renders 1`] = `
<Fragment>
  <Blueprint3.Popover
    boundary="scrollParent"
    captureDismiss={false}
    content={
      <Blueprint3.Menu>
        <Blueprint3.MenuItem
          disabled={true}
          icon="eye-open"
          id="main.js"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="Main Process (main.js)"
        />
        <React.Fragment>
          <Blueprint3.MenuDivider />
          <Blueprint3.MenuItem
            disabled={false}
            icon="plus"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Add Editor"
          />
        </React.Fragment>
        <React.Fragment>
          <Blueprint3.MenuDivider />
          <Blueprint3.MenuItem
            disabled={false}
            icon="grid-view"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Reset Layout"
          />
        </React.Fragment>
      </Blueprint3.Menu>
    }
    defaultIsOpen={false}
    disabled={false}
    fill={false}
    hasBackdrop={false}
    hoverCloseDelay={300}
    hoverOpenDelay={150}
    inheritDarkTheme={true}
    interactionKind="click"
    minimal={false}
    modifiers={Object {}}
    openOnTargetFocus={true}
    position="bottom"
    targetTagName="span"
    transitionDuration={300}
    usePortal={true}
    wrapperTagName="span"
  >
    <Blueprint3.Button
      icon="applications"
      text="Editors"
    />
  </Blueprint3.Popover>
</Fragment>
`;
